src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerActivityExtensions.cs

```diff
@@ -64,8 +64,8 @@ public static JaegerSpan ToJaegerSpan(this Activity activity)
                 Tags = PooledList<JaegerTag>.Create(),
             };
 
-            DictionaryEnumerator<string, string, TagState>.AllocationFreeForEach(
-                activity.Tags,
+            DictionaryEnumerator<string, object, TagState>.AllocationFreeForEach(
+                activity.TagObjects,
```

Comment:
@tarekgh Just noticing Activity.TagObjects uses a custom LinkedList and not ActivityTagsCollection like ActivityEvent & ActivityLink use. Because of that, no struct enumerator, perf regression for us anytime we enumerate TagObjects. Is there a compelling reason for the difference? What we really need is for _whatever_ is underneath the IEnumerable returned to have a `struct GetEnumerator()` operation so we can foreach without allocation.

Something like...

```csharp
   private class TagsLinkedList : IEnumerable<KeyValuePair<string, object>>
   {
        public Enumerator<KeyValuePair<string, object>> GetEnumerator() { return Enumerator(this); }
        IEnumerator<T> IEnumerable<KeyValuePair<string, object>>.GetEnumerator() { return Enumerator(this); }
        IEnumerator IEnumerable.GetEnumerator() { return Enumerator(this); }

        private struct Enumerator<KeyValuePair<string, object>> { ... }
   }

        public IEnumerable<KeyValuePair<string, object?>> TagObjects
        {
            get => _tags ?? Unsafe.As<IEnumerable<KeyValuePair<string, object?>>>(s_emptyBaggageTags);
        }
```

Reply:
This is something can look at to optimize as everything is implementation details. I didn't change anything from what we used to have. Note that this will be a little bit tricky because it is possible some tags get removed while someone else enumerating the list.

Reply:
Awesome this would help a lot. When I say perf regression I mean moving from our old Span object to using Activity to drive everything.

Reply:
If you are interested, you may submit a PR for that? we can work out the details together if you want.

CC @noahfalk

Reply:
@tarekgh Done https://github.com/dotnet/runtime/pull/40362